### DISCUSSION TITLE
Listelemente mit Zustand hinzufügen funktioniert nicht

### DISCUSSION THREAD

**Discussion by AnH13-6** [neuefische/questions#148] [Web] [answered]
Hallo zusammen, 

ich habe gerade versucht anhand eines Tutorials Zustand in meiner App zu implementieren. VS Code zeigt keine Fehler an und der Browser sagt "Encountered two children with the same key, `{vacation.id}`". Könnte das der Grund sein, weshalb ich keine Elemente zu meiner Liste hinzufügen kann? Leider weiß ich auch nicht, wo ich die `{vacation.id}` durch etwas anderes ersetzen könnte.

Ich möchte eine Liste von Vacations bearbeiten, in der ich neue Vacations hinzufügen und bestehende löschen kann. 

Hier der Link zum pull-request des entsprechenden Branchs: https://github.com/AnH13-6/capstone-project/pull/6

Außerdem hier direkt mein Code aus der useVacationStore.js: 
```javascript
import {nanoid} from 'nanoid';
import create from 'zustand';
import {devtools, persist} from 'zustand/middleware';

const vacationStore = set => ({
	vacations: [
		{
			vacationName: 'Italy 2021',
			id: nanoid(),
			entries: [],
		},
		{
			vacationName: 'Japan 2015',
			id: nanoid(),
			entries: [],
		},
	],
	addVacation: vacation => {
		set(state => ({
			vacations: [vacation, ...state.vacations],
		}));
	},
	removeVacation: vacationId => {
		set(state => ({
			vacations: state.vacations.filter(vacation => vacation.id !== vacationId),
		}));
	},
});

const useVacationStore = create(
	devtools(
		persist(vacationStore, {
			name: 'vacations',
		})
	)
);

export default useVacationStore;
```

Dazu gibt es noch die VacationForm.js:
```javascript
import {nanoid} from 'nanoid';
import {useState} from 'react';

import useVacationStore from '../hooks/useVacationStore';

export default function VacationForm() {
	const addVacation = useVacationStore(state => state.addVacation);

	const [vacationName, setVacationName] = useState('');

	const handleVacationAdd = () => {
		addVacation({id: nanoid(), vacationName: vacationName, entries: []});
		setVacationName('');
	};

	return (
		<>
			<form>
				<label>Add a new vacation: </label>
				<input
					value={vacationName}
					onChange={event => {
						event.preventDefault();
						setVacationName(event.target.value);
					}}
					type="text"
					name="newVacation"
					placeholder="ex. Sweden 2022"
					required
				/>
				<button type="submit" onClick={() => handleVacationAdd}>
					Add
				</button>
			</form>
		</>
	);
}

```

Und hier die Vacations.js, in der die Elemete angezeigt werden sollen. Die beiden bestehenden vacations aus der Liste werden angezeigt, aber neue hinzufügen funktioniert nicht.
```javascript
import {HiOutlineBookOpen} from 'react-icons/hi';

import useVacationStore from '../hooks/useVacationStore';

function Vacations() {
	const {vacations, removeVacation} = useVacationStore(state => ({
		vacations: state.vacations,
		removeVacation: state.removeVacation,
	}));
	return (
		<>
			<ul>
				{vacations.map(vacation => (
					<li key={vacation.id}>
						<button>
							{vacation.vacationName}
							<HiOutlineBookOpen />
						</button>
						<button
							onClick={() => {
								removeVacation(vacation.id);
							}}
						>
							Delete
						</button>
					</li>
				))}
			</ul>
		</>
	);
}

export default Vacations;
```

Entschuldigt bitte den langen Post, ich hoffe sehr, dass mich jemand aufklären kann.

**Comment by actyvystom** [comment]
Hallo @AnH13-6 , kannst Du bitte einen Link zu dem zugehörigen Branch in Deinem Repository hinzufügen? Das macht die Fehlersuche einfacher... und noch ein Tipp für das Markdown: wenn Du einen Codeblock verwendest, kannst Du mit Angabe der verwendeten Sprache das Syntax-Highlighting für den jeweiligen Block vorgeben, das erleichtert die Lesbarkeit - also z.B.: 
<pre>
```javascript
import {HiOutlineBookOpen} from 'react-icons/hi';
import useVacationStore from '../hooks/useVacationStore';
```
</pre>

das ergibt dann: 
```javascript
import {HiOutlineBookOpen} from 'react-icons/hi';
import useVacationStore from '../hooks/useVacationStore';
```

**Comment by AnH13-6** [comment]
Danke Thomas, das markdown hab ich aktualisiert. 
Heißt das, dass ich den jetztigen Stand meiner Dateien auf github pushen muss, damit ich einen Link zum entsprechenden Branch erstellen kann?

**Reply by actyvystom** [reply]
Hallo Anne, ja genau. Pushe deinen Branch auf github und dann kannst du den Link dazu hier posten.

**Comment by actyvystom** [comment, marked as answer]
Hallo Anne,
dein `VacationStore` scheint grundsätzlich korrekt angelegt zu sein, das Problem liegt in der Art und Weise, wie Du darauf zugreifst: 
1. In deiner Vacations-Komponente kannst Du deine Vacations-Daten aus dem useVacationStore()-Hook extrahieren

```javascript
const {vacations} = useVacationStore();
```
Dann kannst Du auch korrekt über das vacations-Array aus dem Store mappen: 

```javascript 
{vacations.map(vacation => (<li key={vacation.id}>
   <button>{vacation.vacationName}<HiOutlineBookOpen /></button>
</li>))}
```  

2. Gleiches gilt für Deine VacationForm-Komponente (`VacationForm.js`) - extrahiere hier einfach nur deine Methode zum Hinzufügen der Daten: 
```javascript
const {addVacation} = useVacationStore();
```

und verwende sie dann wie gehabt in deinem Handler:

```javascript
...
addVacation({id: nanoid(), vacationName: vacationName, entries: []});
...
```

**Reply by AnH13-6** [reply]
Vielen Dank!
